src/v/storage/disk_log_impl.cc

```diff
@@ -2574,9 +2574,29 @@ disk_log_impl::remove_prefix_full_segments(truncate_prefix_config cfg) {
       },
       [this] {
           auto ptr = _segs.front();
-          _segs.pop_front();
-          _probe->add_bytes_prefix_truncated(ptr->file_size());
-          return remove_segment_permanently(ptr, "remove_prefix_full_segments");
+          // evict readers before trying to grab a write lock to prevent
+          // contention
+          return _readers_cache->evict_segment_readers(ptr).then(
+            [this, ptr](readers_cache::range_lock_holder cache_lock) {
+                /**
+                 * If segment has outstanding locks wait for it to be unlocked
+                 * before prefixing truncating it, this way the prefix
+                 * truncation will not overlap with appends.
+                 */
+                return ptr->write_lock().then(
+                  [this, ptr, cache_lock = std::move(cache_lock)](
+                    ss::rwlock::holder lock_holder) {
+                      _segs.pop_front();
```

Comment:
Prior to this change, `ptr` and the segment popped are the same.

```
          auto ptr = _segs.front();
          _segs.pop_front();
```

But now, there are two interleaved scheduling points between setting `ptr`, and popping the segment. What guarantees that they are still the same? Does it matter if they are different?